designs/NuGet-PackageFileService.md

```diff
@@ -0,0 +1,104 @@
+
+Package Icons can either come from:
+- a package feeds search service
+- (old style) or a URI to an arbitrary place on the web (specified in the NuSpec via now deprecated [iconUrl](https://docs.microsoft.com/en-us/nuget/reference/nuspec#iconurl) element.)
+- embedded in a package (specified in the NuSpec via [icon](https://docs.microsoft.com/en-us/nuget/reference/nuspec#icon) element.)
+
+Licenses can also come from a few places:
+- a package feeds search service
+- a [licenseUrl](https://docs.microsoft.com/en-us/nuget/reference/nuspec#licenseurl) (deprecated)
+- a [license expression](https://docs.microsoft.com/en-us/nuget/reference/nuspec#license)
+- an [embedded license file](https://docs.microsoft.com/en-us/nuget/reference/nuspec#license) in the package
+
+## Who are the customers
+
+All PM UI customers.
+
+## Goals
+
+Support Codespaces-connected and standalone fetching of icons and licenses.
+
+## Non-Goals
+
+## Solution Overview
+
+### Fetching Files
+Defined a new interface for this service:
+
+```C#
+namespace NuGet.VisualStudio.Internal.Contracts
+{
+    public interface INuGetPackageFileService : IDisposable
+    {
+        ValueTask<Stream?> GetPackageIconAsync(PackageIdentity packageIdentity, CancellationToken cancellationToken);
+        ValueTask<Stream?> GetEmbeddedLicenseAsync(PackageIdentity packageIdentity, CancellationToken cancellationToken);
+    }
+}
+```
+
+On the service side, in [SearchObject.CacheBackgroundData()](https://github.com/NuGet/NuGet.Client/blob/b5b44526dea0379ebd6c8e51e8a041c06d5845ca/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs#L218-L236), beyond the packageSearchMetadata caching that was already done, we call a NuGetPackageFileService instance to AddIconToCache() and to AddLicenseToCache(). These calls cause a PackageIdentity key and a URI value to be stored in a MemoryCache.
+
+On the client side, when trying to fetch the icon, we call PackageFileService.GetPackageIconAsync, passing in the package identity.
+
+On the client side, when trying to fetch an embedded license, we call PackageFileService.GetEmbeddedLicenseAsync.
+
+On the service side, we then retrieve the appropriate file from the appropriate location based on the Uri stored in the MemoryCache and return the stream to the caller.
+
+#### IdentityToUri Memory Cache notes
+
+The settings of the IdentityToUri memory cache that the NuGetPackageFileService class hosts and maintains are identical to the metadata MemoryCache that the search service uses as part of it codespaces rebuilding:
+- 4 MB memory limit
+- 0 physical memory limit percentage
+- 2min polling interval
+
+Given those settings, we don't expect to frequently not find an entry that should be available in the memory cache...we have a Telemetry call to record how often this does happen.

```

Comment:
I had to read this a couple times.  I think the "... don't ... not ..." tripped me up.  Reword for clarity.  Suggestion:

We expect that failures to find an entry in the memory cache due to eviction will be rare, and telemetry will be added to verify this.

Reply:
fixed in next iteration.